File: frontend/src/services/arxivApi.ts
import axios from 'axios'
import type { ArxivAuthor, ArxivPaper, ArxivSearchParams } from '@/features/arxiv/types'

class ArxivApiService {
  private readonly baseUrl = 'https://export.arxiv.org/api/query'

  /**
   * Search arXiv papers by various criteria
   */
  async searchPapers(params: ArxivSearchParams): Promise<ArxivPaper[]> {
    const searchParams = new URLSearchParams()
    
    if (params.query) {
      searchParams.append('search_query', params.query)
    }
    
    if (params.idList && params.idList.length > 0) {
      searchParams.append('id_list', params.idList.join(','))
    }
    
    if (params.start !== undefined) {
      searchParams.append('start', params.start.toString())
    }
    
    if (params.maxResults !== undefined) {
      searchParams.append('max_results', params.maxResults.toString())
    }
    
    if (params.sortBy) {
      searchParams.append('sortBy', params.sortBy)
    }
    
    if (params.sortOrder) {
      searchParams.append('sortOrder', params.sortOrder)
    }

    try {
      const response = await axios.get(`${this.baseUrl}?${searchParams.toString()}`, {
        headers: {
          'Accept': 'application/atom+xml'
        }
      })
      
      return this.parseArxivResponse(response.data)
    } catch (error) {
      console.error('Error fetching from arXiv:', error)
      throw new Error('Failed to fetch papers from arXiv')
    }
  }

  /**
   * Get a paper by arXiv ID
   */
  async getPaperById(arxivId: string): Promise<ArxivPaper | null> {
    const papers = await this.searchPapers({ idList: [arxivId] })
    return papers.length > 0 ? papers[0] : null
  }

  /**
   * Search papers by title
   */
  async searchByTitle(title: string, maxResults = 10): Promise<ArxivPaper[]> {
    return this.searchPapers({
      query: `ti:"${title}"`,
      maxResults,
      sortBy: 'relevance'
    })
  }

  /**
   * Search papers by author
   */
  async searchByAuthor(authorName: string, maxResults = 20): Promise<ArxivPaper[]> {
    return this.searchPapers({
      query: `au:"${authorName}"`,
      maxResults,
      sortBy: 'lastUpdatedDate',
      sortOrder: 'descending'
    })
  }

  /**
   * Search papers by category
   */
  async searchByCategory(category: string, maxResults = 20): Promise<ArxivPaper[]> {
    return this.searchPapers({
      query: `cat:${category}`,
      maxResults,
      sortBy: 'lastUpdatedDate',
      sortOrder: 'descending'
    })
  }

  /**
   * Advanced search with multiple criteria
   */
  async advancedSearch({
    title,
    author,
    abstract,
    category,
    maxResults = 20
  }: {
    title?: string
    author?: string
    abstract?: string
    category?: string
    maxResults?: number
  }): Promise<ArxivPaper[]> {
    const queryParts: string[] = []
    
    if (title) queryParts.push(`ti:"${title}"`)
    if (author) queryParts.push(`au:"${author}"`)
    if (abstract) queryParts.push(`abs:"${abstract}"`)
    if (category) queryParts.push(`cat:${category}`)
    
    if (queryParts.length === 0) {
      throw new Error('At least one search criterion must be provided')
    }
    
    const query = queryParts.join(' AND ')
    
    return this.searchPapers({
      query,
      maxResults,
      sortBy: 'relevance'
    })
  }

  /**
   * Parse arXiv XML response to structured data
   */
  private parseArxivResponse(xmlData: string): ArxivPaper[] {
    const parser = new DOMParser()
    const doc = parser.parseFromString(xmlData, 'text/xml')
    const entries = doc.querySelectorAll('entry')
    
    const papers: ArxivPaper[] = []
    
    entries.forEach(entry => {
      try {
        const paper = this.parseEntry(entry)
        if (paper) papers.push(paper)
      } catch (error) {
        console.warn('Error parsing arXiv entry:', error)
      }
    })
    
    return papers
  }

  private parseEntry(entry: Element): ArxivPaper | null {
    const getId = (text: string) => text.split('/').pop()?.replace('v', '') || text
    
    const id = getId(entry.querySelector('id')?.textContent || '')
    const title = entry.querySelector('title')?.textContent?.trim().replace(/\s+/g, ' ') || ''
    const summary = entry.querySelector('summary')?.textContent?.trim().replace(/\s+/g, ' ') || ''
    const published = entry.querySelector('published')?.textContent || ''
    const updated = entry.querySelector('updated')?.textContent || ''
    
    // Parse authors
    const authorElements = entry.querySelectorAll('author')
    const authors: ArxivAuthor[] = []
    authorElements.forEach(authorEl => {
      const name = authorEl.querySelector('name')?.textContent?.trim()
      const affiliation = authorEl.querySelector('affiliation')?.textContent?.trim()
      if (name) {
        authors.push({ name, affiliation })
      }
    })
    
    // Parse categories
    const categoryElements = entry.querySelectorAll('category')
    const categories: string[] = []
    let primaryCategory = ''
    
    categoryElements.forEach((catEl, index) => {
      const term = catEl.getAttribute('term')
      if (term) {
        categories.push(term)
        if (index === 0) primaryCategory = term
      }
    })
    
    // Parse links
    const linkElements = entry.querySelectorAll('link')
    let pdfUrl = ''
    let abstractUrl = ''
    
    linkElements.forEach(linkEl => {
      const href = linkEl.getAttribute('href') || ''
      const type = linkEl.getAttribute('type')
      const title = linkEl.getAttribute('title')
      
      if (type === 'application/pdf' || title === 'pdf') {
        pdfUrl = href
      } else if (href.includes('abs/')) {
        abstractUrl = href
      }
    })
    
    // Parse optional fields
    const doi = entry.querySelector('arxiv\\:doi, doi')?.textContent?.trim()
    const journalRef = entry.querySelector('arxiv\\:journal_ref, journal_ref')?.textContent?.trim()
    const comment = entry.querySelector('arxiv\\:comment, comment')?.textContent?.trim()
    
    if (!id || !title) return null
    
    return {
      id,
      title,
      summary,
      authors,
      published,
      updated,
      categories,
      primaryCategory,
      pdfUrl,
      abstractUrl,
      doi,
      journalRef,
      comment
    }
  }
}

export const arxivApi = new ArxivApiService()

File: backend/app/services/arxiv_api.py
"""
ArXiv API service for DeepCite
"""
import logging
import httpx
from typing import List, Dict, Any, Optional
from xml.etree import ElementTree as ET

logger = logging.getLogger(__name__)


class ArxivAuthor:
    def __init__(self, name: str, affiliation: Optional[str] = None):
        self.name = name
        self.affiliation = affiliation


class ArxivPaper:
    def __init__(self, **kwargs):
        self.id = kwargs.get('id', '')
        self.title = kwargs.get('title', '')
        self.summary = kwargs.get('summary', '')
        self.authors = kwargs.get('authors', [])
        self.published = kwargs.get('published', '')
        self.updated = kwargs.get('updated', '')
        self.categories = kwargs.get('categories', [])
        self.primaryCategory = kwargs.get('primaryCategory', '')
        self.pdfUrl = kwargs.get('pdfUrl', '')
        self.abstractUrl = kwargs.get('abstractUrl', '')
        self.doi = kwargs.get('doi')
        self.journalRef = kwargs.get('journalRef')
        self.comment = kwargs.get('comment')


class ArxivApiService:
    """Service for interacting with arXiv API"""

    def __init__(self):
        self.base_url = 'https://export.arxiv.org/api/query'

    async def searchPapers(self, params: Dict[str, Any]) -> List[ArxivPaper]:
        """Search arXiv papers by various criteria"""
        search_params = {}

        if 'query' in params:
            search_params['search_query'] = params['query']

        if 'idList' in params and params['idList']:
            search_params['id_list'] = ','.join(params['idList'])

        if 'start' in params:
            search_params['start'] = str(params['start'])

        if 'maxResults' in params:
            search_params['max_results'] = str(params['maxResults'])

        if 'sortBy' in params:
            search_params['sortBy'] = params['sortBy']

        if 'sortOrder' in params:
            search_params['sortOrder'] = params['sortOrder']

        try:
            async with httpx.AsyncClient(timeout=30.0) as client:
                response = await client.get(
                    self.base_url,
                    params=search_params,
                    headers={'Accept': 'application/atom+xml'}
                )

                if response.status_code != 200:
                    logger.error(f"arXiv API error: {response.status_code}")
                    return []

                return self._parse_arxiv_response(response.text)

        except Exception as e:
            logger.error(f"Error fetching from arXiv: {e}")
            return []

    def _parse_arxiv_response(self, xml_data: str) -> List[ArxivPaper]:
        """Parse arXiv XML response to structured data"""
        try:
            root = ET.fromstring(xml_data)
            entries = root.findall('.//{http://www.w3.org/2005/Atom}entry')

            papers = []
            for entry in entries:
                paper = self._parse_entry(entry)
                if paper:
                    papers.append(paper)

            return papers

        except Exception as e:
            logger.error(f"Error parsing arXiv response: {e}")
            return []

    def _parse_entry(self, entry) -> Optional[ArxivPaper]:
        """Parse individual arXiv entry"""
        try:
            # Extract basic fields
            id_elem = entry.find('.//{http://www.w3.org/2005/Atom}id')
            title_elem = entry.find('.//{http://www.w3.org/2005/Atom}title')
            summary_elem = entry.find('.//{http://www.w3.org/2005/Atom}summary')
            published_elem = entry.find('.//{http://www.w3.org/2005/Atom}published')
            updated_elem = entry.find('.//{http://www.w3.org/2005/Atom}updated')

            if not id_elem or not title_elem:
                return None

            # Extract ID from URL
            id_url = id_elem.text or ''
            paper_id = id_url.split('/')[-1] if '/' in id_url else id_url

            title = (title_elem.text or '').strip()
            summary = (summary_elem.text or '').strip()
            published = published_elem.text or '' if published_elem else ''
            updated = updated_elem.text or '' if updated_elem else ''

            # Parse authors
            authors = []
            author_elements = entry.findall('.//{http://www.w3.org/2005/Atom}author')
            for author_elem in author_elements:
                name_elem = author_elem.find('.//{http://www.w3.org/2005/Atom}name')
                if name_elem is not None and name_elem.text:
                    name = name_elem.text.strip()
                    # Try to get affiliation
                    affiliation = None
                    affiliation_elem = author_elem.find('.//{http://arxiv.org/schemas/atom}affiliation')
                    if affiliation_elem is not None:
                        affiliation = affiliation_elem.text.strip()
                    authors.append(ArxivAuthor(name, affiliation))

            # Parse categories
            categories = []
            primary_category = ''
            category_elements = entry.findall('.//{http://www.w3.org/2005/Atom}category')
            for i, cat_elem in enumerate(category_elements):
                term = cat_elem.get('term', '')
                if term:
                    categories.append(term)
                    if i == 0:  # First category is primary
                        primary_category = term

            # Parse links
            pdf_url = ''
            abstract_url = ''
            link_elements = entry.findall('.//{http://www.w3.org/2005/Atom}link')
            for link_elem in link_elements:
                href = link_elem.get('href', '')
                rel = link_elem.get('rel', '')
                type_attr = link_elem.get('type', '')
                title_attr = link_elem.get('title', '')

                if type_attr == 'application/pdf' or title_attr == 'pdf':
                    pdf_url = href
                elif 'abs' in href:
                    abstract_url = href

            # Optional fields
            doi = None
            journal_ref = None
            comment = None

            doi_elem = entry.find('.//{http://arxiv.org/schemas/atom}doi')
            if doi_elem is not None:
                doi = doi_elem.text.strip()

            journal_elem = entry.find('.//{http://arxiv.org/schemas/atom}journal_ref')
            if journal_elem is not None:
                journal_ref = journal_elem.text.strip()

            comment_elem = entry.find('.//{http://arxiv.org/schemas/atom}comment')
            if comment_elem is not None:
                comment = comment_elem.text.strip()

            return ArxivPaper(
                id=paper_id,
                title=title,
                summary=summary,
                authors=authors,
                published=published,
                updated=updated,
                categories=categories,
                primaryCategory=primary_category,
                pdfUrl=pdf_url,
                abstractUrl=abstract_url,
                doi=doi,
                journalRef=journal_ref,
                comment=comment
            )

        except Exception as e:
            logger.warning(f"Error parsing arXiv entry: {e}")
            return None
